### PR TITLE
fix: Fuzzer fixing

### PR DESCRIPTION
**Pull request by asmelko**
This PR fixes major bugs found by the parser library fuzzer.

### CODE REVIEW

File: parser_library/CMakeLists.txt [hunk 1]
```diff
@@ -250,6 +250,11 @@ if(BUILD_FUZZER)
 
     set_target_properties(library_fuzzer PROPERTIES LINK_FLAGS "-fsanitize=fuzzer")
 	set_target_properties(library_fuzzer PROPERTIES COMPILE_FLAGS "-DANTLR4CPP_STATIC")
+
```
Comment (michalbali256) [inline parser_library/CMakeLists.txt]: Is the file copied properly every time it is changed?

File: parser_library/fuzzer.dict [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+"ORG"
+"L'"
+"I'"
+"S'"
+"T'"
+"D'"
+"O'"
+"N'"
+"K'"
+"C'ABC'"
```
Comment (michalbali256) [inline parser_library/fuzzer.dict]: Have you thought about adding some machine instructions as well?

File: parser_library/CMakeLists.txt (hunk 1 above)
Comment (asmelko) [reply parser_library/CMakeLists.txt]: If the input file is modified the build system will re-run CMake to re-configure the file and generate the build system again.

File: parser_library/fuzzer.dict (hunk 2 above)
Comment (asmelko) [reply parser_library/fuzzer.dict]: Yes, I will add more extensive dictionary in future commits.

File: parser_library/src/context/ordinary_assembly/dependency_collector.cpp [hunk 3]
```diff
@@ -121,14 +121,16 @@ dependency_collector& dependency_collector::div_mul(const dependency_collector&
...
+	auto unknown_space = std::find_if(addr.spaces.begin(), addr.spaces.end(), [](auto& entry) { return entry.first->kind == context::space_kind::LOCTR_UNKNOWN; });
 	if (unknown_space != addr.spaces.end())
 	{
-		for (auto it = addr.spaces.begin(); it != addr.spaces.end(); ++it)
-			if (it->first->kind != context::space_kind::LOCTR_UNKNOWN)
-				it->first->remove_listener(&addr);
-		auto entry = *unknown_space;
-		addr.spaces.clear();
-		addr.spaces.push_back(std::move(entry));
+		for (int i = addr.spaces.size() - 1; i > 0; --i)
```
Comment (slavek-kucera) [inline parser_library/src/context/ordinary_assembly/dependency_collector.cpp]: is the zero-th element ignored intentionally here?

Comment (asmelko) [reply parser_library/src/context/ordinary_assembly/dependency_collector.cpp]: no! I will fix it.

**Comment by asmelko** [comment]
It is true that the above mentioned pairs would use static vector but for now I would try to retain the idea that each statement field is optional. I will address this problem in future PRs and will try to get rid of the pairs.

File: parser_library/src/context/ordinary_assembly/dependency_collector.cpp [hunk 4]
```diff
@@ -121,14 +121,16 @@ dependency_collector& dependency_collector::div_mul(const dependency_collector&
 
 void dependency_collector::adjust_address(address& addr)
 {
-	auto unknown_space = std::find_if(addr.spaces.begin(), addr.spaces.end(), [] (auto& entry) { return entry.first->kind == context::space_kind::LOCTR_UNKNOWN; });
+	auto unknown_space = std::find_if(addr.spaces.begin(), addr.spaces.end(), [](auto& entry) { return entry.first->kind == context::space_kind::LOCTR_UNKNOWN; });
```
Comment (slavek-kucera) [inline parser_library/src/context/ordinary_assembly/dependency_collector.cpp:124]: I suspect that the erase inside the loop causes O(N^2) behavior.
```suggestion
	auto known_spaces = std::partition(addr.spaces.begin(), addr.spaces.end(), [](auto& entry) { return entry.first->kind == context::space_kind::LOCTR_UNKNOWN; });
	std::for_each(known_spaces, addr.spaces.end(), [&addr](auto& entry) { return entry.first->remove_listener(&addr); });
	addr.spaces.erase(known_spaces, addr.spaces.end());
```

**Pull request merged**